File: java/src/quinoa/items/Item.java
package quinoa.items;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.util.ArrayList;
import java.util.UUID;
import quinoa.Copyable;
import quinoa.Removable;
import quinoa.Storable;
import quinoa.items.ItemManager.ItemCategory;
import quinoa.items.ItemManager.ItemClass;
import quinoa.items.ItemManager.ItemState;

public class Item implements Removable, Storable, Copyable
{
    private ItemCategory itemCategory;
    private ItemClass itemClass;
    private ItemState itemState;
    private int x, y;
    private String ID;
    private int maxStackSize;
    private int stackSize;
    private double worth;
    private ArrayList<ItemAttribute> attributes;

    public Item()
    {
        x = 0;
        y = 0;
        itemState = ItemState.GROUND;
        itemCategory = ItemCategory.TOOL;
        ID = UUID.randomUUID().toString();
        maxStackSize = 1;
        stackSize = 1;
        worth = 0.01;
        attributes = new ArrayList<ItemAttribute>();
        setItemClass(ItemClass.LOG);
    }

    //Make a new ID for the item
    public void refreshID()
    {
        ID = UUID.randomUUID().toString();
    }

    public void setPosition(int x, int y)
    {
        this.x = x;
        this.y = y;
    }

    public boolean shouldBeRemoved()
    {
        if(getItemState() == ItemState.DESTROYED)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    public void remove()
    {
        setItemState(ItemState.DESTROYED);
    }

    public void save(BufferedWriter out) throws Exception
    {
        out.write(this.getItemCategory().name()); out.newLine();
        out.write(this.getItemClass().name()); out.newLine();
        out.write(this.getItemState().name()); out.newLine();
        out.write(this.getX() + ""); out.newLine();
        out.write(this.getY() + ""); out.newLine();
        out.write(this.getID()); out.newLine();
        out.write(this.maxStackSize + ""); out.newLine();
        out.write(this.stackSize + ""); out.newLine();
        out.write(this.getWorth() + ""); out.newLine();
        out.write(this.getAttributes().size() + ""); out.newLine();
        for(ItemAttribute tempIA : getAttributes())
        {
            tempIA.save(out);
        }
    }

    public void load(BufferedReader in) throws Exception
    {
        this.setItemCategory(getItemCategory().valueOf(in.readLine()));
        this.setItemClass(getItemClass().valueOf(in.readLine()));
        this.setItemState(getItemState().valueOf(in.readLine()));
        this.x = Integer.parseInt(in.readLine());
        this.y = Integer.parseInt(in.readLine());
        this.ID = in.readLine();
        this.maxStackSize = Integer.parseInt(in.readLine());
        this.stackSize = Integer.parseInt(in.readLine());
        this.setWorth(Double.parseDouble(in.readLine()));
        attributes = new ArrayList<ItemAttribute>();
        int attributeSize = Integer.parseInt(in.readLine());
        for(int i=0; i < attributeSize; i++)
        {
            ItemAttribute newIA = new ItemAttribute();
            newIA.setType(ItemAttribute.ItemAttributeType.valueOf(in.readLine()));
            newIA.setParameter(in.readLine());
            getAttributes().add(newIA);
        }
    }

    public Object copy()
    {
        Item newItem = new Item();
        newItem.setItemCategory(this.getItemCategory());
        newItem.setItemClass(this.getItemClass());
        newItem.setItemState(this.getItemState());
        newItem.x = this.getX();
        newItem.y = this.getY();
        newItem.ID = this.getID();
        newItem.maxStackSize = this.getMaxStackSize();
        newItem.stackSize = this.getStackSize();
        newItem.setWorth(this.getWorth());
        newItem.getAttributes().clear();
        for(ItemAttribute tempIA : this.getAttributes())
        {
            newItem.getAttributes().add((ItemAttribute)tempIA.copy());
        }
        return newItem;
    }

    /**
     * @return the itemCategory
     */
    public ItemCategory getItemCategory() {
        return itemCategory;
    }

    /**
     * @return the itemClass
     */
    public ItemClass getItemClass() {
        return itemClass;
    }

    /**
     * @return the itemState
     */
    public ItemState getItemState() {
        return itemState;
    }

    /**
     * @param itemState the itemState to set
     */
    public void setItemState(ItemState itemState) {
        this.itemState = itemState;
    }

    /**
     * @return the x
     */
    public int getX() {
        return x;
    }

    /**
     * @return the y
     */
    public int getY() {
        return y;
    }

    /**
     * @return the ID
     */
    public String getID() {
        return ID;
    }

    /**
     * @param itemCategory the itemCategory to set
     */
    public void setItemCategory(ItemCategory itemCategory) {
        this.itemCategory = itemCategory;
    }

    /**
     * @param itemClass the itemClass to set
     */
    public void setItemClass(ItemClass itemClass)
    {
        this.itemClass = itemClass;
        ItemManager.initialize(this);
    }

    /**
     * @return the maxStackSize
     */
    public int getMaxStackSize() {
        return maxStackSize;
    }

    /**
     * @param maxStackSize the maxStackSize to set
     */
    public void setMaxStackSize(int maxStackSize) {
        this.maxStackSize = maxStackSize;
    }

    /**
     * @return the stackSize
     */
    public int getStackSize() {
        return stackSize;
    }

    /**
     * @param stackSize the stackSize to set
     */
    public void setStackSize(int stackSize) {
        this.stackSize = stackSize;
    }

    /**
     * @return the attributes
     */
    public ArrayList<ItemAttribute> getAttributes() {
        return attributes;
    }

    /**
     * @return the worth
     */
    public double getWorth() {
        return worth;
    }

    /**
     * @param worth the worth to set
     */
    public void setWorth(double worth) {
        this.worth = worth;
    }
}

File: dotnet/Core/Items/Item.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using LumberjackRL.Core.Utilities;
using System.IO;

namespace LumberjackRL.Core.Items
{
    public class Item : IRemoveObject, IStoreObject, ICopyObject
    {
        private ItemClass m_itemClass;

        public ItemCategory itemCategory
        {
            get;
            set;
        }

        public ItemClass itemClass
        {
            get
            {
                return m_itemClass;
            }

            set
            {
                m_itemClass = value;
                ItemManager.initialize(this);
            }
        }

        public ItemState itemState
        {
            get;
            set;
        }
        
        public bool ShouldBeRemoved
        {
            get
            {
                return (itemState == ItemState.DESTROYED);
            }
        }

        public int x
        {
            get;
            set;
        }
        
        public int y
        {
            get;
            set;
        }
        
        public String ID
        {
            get;
            set;
        }
        
        public int maxStackSize
        {
            get;
            set;
        }
        
        public int stackSize
        {
            get;
            set;
        }
        
        public double worth
        {
            get;
            set;
        }

        public List<ItemAttribute> attributes
        {
            get;
            set;
        }

        public Item()
        {
            x = 0;
            y = 0;
            itemState = ItemState.GROUND;
            itemCategory = ItemCategory.TOOL;
            ID = RandomNumber.RandomUUID().ToString();
            maxStackSize = 1;
            stackSize = 1;
            worth = 0.01;
            attributes = new List<ItemAttribute>();
            itemClass = ItemClass.LOG;
        }

        /// <summary>
        /// Make a new ID for the item.
        /// </summary>
        public void refreshID()
        {
            ID = RandomNumber.RandomUUID().ToString();
        }

        public void setPosition(int x, int y)
        {
            this.x = x;
            this.y = y;
        }
        
        public void RemoveObject()
        {
            itemState = ItemState.DESTROYED;
        }

        public void SaveObject(StreamWriter outStream)
        {
            outStream.WriteLine(itemCategory.ToString());
            outStream.WriteLine(itemClass.ToString());
            outStream.WriteLine(itemState.ToString());
            outStream.WriteLine(this.x + "");
            outStream.WriteLine(this.y + "");
            outStream.WriteLine(this.ID);
            outStream.WriteLine(this.maxStackSize + "");
            outStream.WriteLine(this.stackSize + "");
            outStream.WriteLine(this.worth + "");
            outStream.WriteLine(this.attributes.Count + "");
            foreach(ItemAttribute tempIA in attributes)
            {
                tempIA.SaveObject(outStream);
            }
        }

        public void LoadObject(StreamReader inStream)
        {
            this.itemCategory = (ItemCategory)Enum.Parse(typeof(ItemCategory), inStream.ReadLine());
            this.itemClass = (ItemClass)Enum.Parse(typeof(ItemClass), inStream.ReadLine());
            this.itemState = (ItemState)Enum.Parse(typeof(ItemState), inStream.ReadLine());
            this.x = Int32.Parse(inStream.ReadLine());
            this.y = Int32.Parse(inStream.ReadLine());
            this.ID = inStream.ReadLine();
            this.maxStackSize = Int32.Parse(inStream.ReadLine());
            this.stackSize = Int32.Parse(inStream.ReadLine());
            this.worth = Double.Parse(inStream.ReadLine());
            attributes = new List<ItemAttribute>();
            int attributeSize = Int32.Parse(inStream.ReadLine());
            for (int i = 0; i < attributeSize; i++)
            {
                ItemAttribute newIA = new ItemAttribute();
                newIA.type = (ItemAttributeType)Enum.Parse(typeof(ItemAttributeType), inStream.ReadLine());
                newIA.parameter = inStream.ReadLine();
                attributes.Add(newIA);
            }
        }

        public Object CopyObject()
        {
            Item newItem = new Item();
            newItem.itemCategory = itemCategory;
            newItem.itemClass = itemClass;
            newItem.itemState = itemState;
            newItem.x = x;
            newItem.y = y;
            newItem.ID = ID;
            newItem.maxStackSize = maxStackSize;
            newItem.stackSize = stackSize;
            newItem.worth = worth;
            newItem.attributes.Clear();
            foreach(ItemAttribute tempIA in attributes)
            {
                newItem.attributes.Add((ItemAttribute)tempIA.CopyObject());
            }
            return newItem;
        }
    }
}
